src/backend/distributed/executor/subplan_execution.c

```diff
@@ -81,3 +117,148 @@ ExecuteSubPlans(DistributedPlan *distributedPlan)
 		FreeExecutorState(estate);
 	}
 }
+
+
+/*
+ * RecordSubplanExecutionsOnNodes iterates over the usedSubPlanNodeList,
+ * and for each entry, record the workerNodes that are accessed by
+ * the distributed plan.
+ *
+ * Later, we'll use this information while we broadcast the intermediate
+ * results to the worker nodes. The idea is that the intermediate result
+ * should only be broadcasted to the worker nodes that are accessed by
+ * the distributedPlan(s) that the subPlan is used in.
+ *
+ * Finally, the function recursively descends into the actual subplans
+ * of the input distributedPlan as well.
+ */
+static void
+RecordSubplanExecutionsOnNodes(HTAB *intermediateResultHash,
+							   DistributedPlan *distributedPlan)
+{
+	List *usedSubPlanNodeList = distributedPlan->usedSubPlanNodeList;
+	ListCell *usedSubPlanNodeCell = NULL;
+	List *subPlanList = distributedPlan->subPlanList;
+	ListCell *subPlanCell = NULL;
+
+	foreach(usedSubPlanNodeCell, usedSubPlanNodeList)
+	{
+		Const *resultIdConst = lfirst(usedSubPlanNodeCell);
+		Datum resultIdDatum = resultIdConst->constvalue;
+		char *resultId = TextDatumGetCString(resultIdDatum);
+
+		bool found = false;
+		IntermediateResultHashEntry *entry = SearchIntermediateResult(
+			intermediateResultHash, resultId, &found);
+
+
+		/* no need to traverse the whole plan if all the workers are hit */
+		if (list_length(entry->nodeList) == GetWorkerNodeCount())
+		{
+			elog(DEBUG4, "subplan %s is used in all workers", resultId);
+
+			break;
+		}
+		else
+		{
+			entry->nodeList = AppendAllAccessedWorkerNodes(entry->nodeList,
+														   distributedPlan);
+
+			elog(DEBUG4, "subplan %s is used in %lu", resultId, distributedPlan->planId);
+		}
+	}
+
+	/* descend into the subPlans */
+	foreach(subPlanCell, subPlanList)
+	{
+		DistributedSubPlan *subPlan = (DistributedSubPlan *) lfirst(subPlanCell);
+		CustomScan *customScan = FetchCitusCustomScanIfExists(subPlan->plan->planTree);
+		if (customScan)
+		{
+			DistributedPlan *distributedPlanOfSubPlan = GetDistributedPlan(customScan);
+			RecordSubplanExecutionsOnNodes(intermediateResultHash,
+										   distributedPlanOfSubPlan);
+		}
+	}
+}
+
+
+/*
+ * AppendAllAccessedWorkerNodes iterates over all the tasks in a distributed plan
+ * to create the list of worker nodes that can be accessed when this plan is executed.
+ *
+ * If there are multiple placements of a Shard, all of them are considered.
+ */
+static List *
+AppendAllAccessedWorkerNodes(List *workerNodeList, DistributedPlan *distributedPlan)
+{
+	List *taskList = distributedPlan->workerJob->taskList;
```

Comment:
Instead of calling `GetWorkerNodeCount()` too many times, maybe set to a variable (here and the other usages as well)

Reply:
No longer calling it multiple times. setting it to a variable and sending it to `AppendAllAccessedWorkerNodes()` as a param